src/compliance_check/src/main.nr

```diff
@@ -1,17 +1,258 @@
+        // }
+
+        let network_epoch = _calculate_epoch(network_timestamp);
+        let user_epoch = _calculate_epoch(timestamp);
+
+        // assert(network_epoch == user_epoch, "epoch is not the same");
+
+        //assert the root hash is valid
+        //change to the correct value
+        assert(root_hash == storage.registry_root.read(), "root hash is not valid");
+
+        //check the zk_id is valid
+        let zk_id_exists = storage.epoch_zk_id.at(user_epoch).at(zk_id).read();
+        assert(zk_id_exists == false, "zk_id has already been registered for this epoch");
+
+        //store the zk_id in the epoch
+        //using user_epoch for network consistency
+        storage.epoch_zk_id.at(user_epoch).at(zk_id).write(true);
+    }
+
+    #[private]
+    fn authorize_transfer_private(_sender: AztecAddress, _from: AztecAddress, _amount: u128) {
+
+        //better way to get the epoch
+
+
+        //check fetch the note for the user
+        let auth_note = storage.address_authorised.at(_sender).get_note().note;
+
+        let note_epoch = auth_note.epoch;
+        let zk_id = auth_note.zk_id;
+
+        //check the note is authenticated
+        assert(auth_note.authenticated == true, "note is not authenticated");
+
+        //check the transfer limit
+        let transfer_total = auth_note.transfer_total + _amount;
+        assert(transfer_total <= TRANSFER_LIMIT, "transfer limit exceeded");
+
+        //update the note for storage
+        let new_auth_note = AuthNote::new(auth_note.owner, zk_id, true, note_epoch, transfer_total);
+
+        storage.address_authorised.at(_sender).replace(new_auth_note).emit(encode_and_encrypt_note(
+            &mut context,
+            _sender,
+            _sender,
+        ));
+        //check the epoch from the note.
+
+        //check the provided timestamp is correct
+        ComplianceCheck::at(context.this_address())._check_epoch_and_zk_id(note_epoch, zk_id).enqueue(&mut context);
+
+    }
+
+    #[public]
+    #[internal]
+    fn _check_epoch_and_zk_id(note_epoch: u64, zk_id: Field) {
+        let network_timestamp = context.timestamp();
+        let note_timestamp = note_epoch * SECONDS_PER_MONTH;
+
+        //comment out for test
+        //  if (note_timestamp >= network_timestamp) {
+        //     assert((network_timestamp + 6000000 as u64) >= note_timestamp, "timestamp is too far in the past");
+        // } else {
+        //     assert((note_timestamp + 6000000 as u64) >= network_timestamp, "timestamp is too far in the future");
+        // }
+
+        let network_epoch = _calculate_epoch(network_timestamp);
+        // assert(network_epoch == note_epoch, "epoch is not the same");
+
+        //check the zk_id has been registered
+        let zk_id_exists = storage.epoch_zk_id.at(note_epoch).at(zk_id).read();
+        assert(zk_id_exists == true, "zk_id has not been registered for this epoch");
+    }
+
+    //Not needed right?
```

Comment:
let's ask the community about this.
i'm thinking the following, we can do as UniV4 hooks mechanism, in which if the Transfer Authority address starts with a `11` bits, we check every transfer, if it's `10` we check private only, `01` public only, and `00` idk